routing/index_router.cpp

```diff
@@ -260,16 +273,131 @@ IRouter::ResultCode IndexRouter::DoCalculateRoute(string const & startCountry,
 
   CHECK_GREATER_OR_EQUAL(segments.size(), routingResult.path.size(), ());
 
-  if (!RedressRoute(segments, delegate, forSingleMwm, starter, route))
-    return IRouter::InternalError;
-  if (delegate.IsCancelled())
-    return IRouter::Cancelled;
+  auto redressResult = RedressRoute(segments, delegate, forSingleMwm, starter, route);
+  if (redressResult != IRouter::NoError)
+    return redressResult;
+
+  m_lastRoute.Init(startPoint, finalPoint);
+  for (Segment const & segment : routingResult.path)
+    m_lastRoute.AddStep(segment, starter.GetPoint(segment, true /* front */));
+
+  return IRouter::NoError;
+}
+
+IRouter::ResultCode IndexRouter::AdjustRoute(platform::CountryFile const & startCountry,
+                                             m2::PointD const & startPoint,
+                                             m2::PointD const & startDirection,
+                                             m2::PointD const & finalPoint,
+                                             double propagateDistance,
+                                             RouterDelegate const & delegate, Route & route)
+{
+  my::Timer timer;
+  TrafficStash::Guard guard(*m_trafficStash);
+  WorldGraph graph = MakeWorldGraph();
+  graph.SetMode(WorldGraph::Mode::NoLeaps);
+
+  Segment startSegment;
+  if (!FindClosestSegment(startCountry, startPoint, true /* isOutgoing */, graph, startSegment))
+    return IRouter::StartPointNotFound;
+
+  IndexGraphStarter starter(
+      IndexGraphStarter::FakeVertex(startSegment, startPoint),
+      IndexGraphStarter::FakeVertex(m_lastRoute.GetFinishSegment(), finalPoint), graph);
+
+  AStarProgress progress(0, 100);
+  progress.Initialize(starter.GetStartVertex().GetPoint(), starter.GetFinishVertex().GetPoint());
+
+  uint32_t drawPointsStep = 0;
+  auto onVisitJunction = [&](Segment const & from, Segment const & /* to */) {
+    m2::PointD const & point = starter.GetPoint(from, true /* front */);
+    auto const lastValue = progress.GetLastValue();
+    auto const newValue = progress.GetProgressForDirectedAlgo(point);
+    if (newValue - lastValue > kProgressInterval)
+      delegate.OnProgress(newValue);
+    if (drawPointsStep % kDrawPointsPeriod == 0)
+      delegate.OnPointCheck(point);
+    ++drawPointsStep;
+  };
+
+  AStarAlgorithm<IndexGraphStarter> algorithm;
+  RoutingResult<Segment> routingResult;
+
+  auto const & steps = m_lastRoute.GetSteps();
+  set<Segment> routeSegmentsSet;
+  for (auto const & step : steps)
+  {
+    auto const & segment = step.GetSegment();
+    if (!IndexGraphStarter::IsFakeSegment(segment))
+      routeSegmentsSet.insert(segment);
+  }
+
+  auto const isFinalVertex = [&](Segment const & vertex) {
```

Comment:
In general, this is wrong, as A-star algorithm assumes that heuristic calculations should be consistent with the set of final vertices. I.e. if, instead of a single final vertex, you have a set of possible final vertices and you just want to find the shortest path to any of them, heuristic should be calculated as:
```c++
double CalcHeurisric(Vertex u) {
  double h = DBL_MAX;
  for (Vertex v : m_targets)
    h = min(h, DistanceOnEarth(u, v));
  return h;
}
```